libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -389,23 +399,55 @@ void ContractCompiler::appendFunctionSelector(ContractDefinition const& _contrac
 		appendInternalSelector(callDataUnpackerEntryPoints, sortedIDs, notFound, m_optimiserSettings.expectedExecutionsPerDeployment);
 	}
 
-	m_context << notFound;
+	m_context << notFoundOrReceiveEther;
 
-	if (fallback)
+	if (!fallback && !etherReceiver)
+		m_context.appendRevert();
+	else
 	{
-		solAssert(!_contract.isLibrary(), "");
-		if (!fallback->isPayable() && needToAddCallvalueCheck)
-			appendCallValueCheck();
+		std::optional<eth::AssemblyItem> receiveEther;
 
-		solAssert(fallback->isFallback(), "");
-		solAssert(FunctionType(*fallback).parameterTypes().empty(), "");
-		solAssert(FunctionType(*fallback).returnParameterTypes().empty(), "");
-		fallback->accept(*this);
-		m_context << Instruction::STOP;
+		if (etherReceiver)
+		{
+			receiveEther = m_context.newTag();
+			// directly jump to receive ether function, if there is no calldata
+			m_context << Instruction::CALLDATASIZE << Instruction::ISZERO;
+			m_context.appendConditionalJumpTo(*receiveEther);
+		}
+
+		m_context << notFound;
+		if (fallback)
+		{
+			solAssert(!_contract.isLibrary(), "");
+			if (!fallback->isPayable() && needToAddCallvalueCheck)
+				appendCallValueCheck();
+
+			solAssert(fallback->isFallback(), "");
+			solAssert(FunctionType(*fallback).parameterTypes().empty(), "");
+			solAssert(FunctionType(*fallback).returnParameterTypes().empty(), "");
+			fallback->accept(*this);
+			m_context << Instruction::STOP;
+		}
+		else
+			// TODO: error message here?
+			m_context.appendRevert();
+
+		if (etherReceiver)
+		{
+			solAssert(receiveEther, "");
+			m_context << *receiveEther;
+			// TODO: should this revert, if value is zero?
+			solAssert(!_contract.isLibrary(), "");
+			solAssert(etherReceiver->isEtherReceiver(), "");
+			solAssert(FunctionType(*etherReceiver).parameterTypes().empty(), "");
+			solAssert(FunctionType(*etherReceiver).returnParameterTypes().empty(), "");
+			etherReceiver->accept(*this);
+			m_context << Instruction::STOP;
+		}
+		else
+			m_context.appendRevert();
```

Comment:
I think this would be unreachable

Reply:
Reachable if ``fallback && !etherReceiver`` - the case ``!fallback && !etherReceiver" above is by the way only separated to avoid increasing gas costs we'd get otherwise.

Reply:
If `!etherReceiver` it just falls through to `fallback`, right?
If `fallback`, it ends with `m_context << Instruction::STOP;`.
If `!fallback`, it ends with `m_context.appendRevert();`.

Reply:
You were right, this is unreachable :-).